pkgs/sage-conf_meson/sage_conf.py

```diff
@@ -0,0 +1,6 @@
+from _sage_conf.__main__ import _main
+
+from builddir._conf import *
+from builddir.build_info import CONDA_PREFIX, SAGE_ROOT
```

Comment:
This "from builddir... import" makes me a bit uncomfortable. 
This seems to assume that `.` is in sys.path
Please test that also non-editable installs (via wheels) of sage-conf_meson work.

Reply:
The `builddir` here is generated under `pkgs/sage-conf_meson`

Reply:
Yes, I knew that when I wrote the above. So?

Reply:
I don't see the difference between `configure` creating the `_conf.py` file and here we are using `meson` to create the `builddir` folder. Can you please be more specific what problem you encounter?

Reply:
> I don't see the difference

In https://github.com/sagemath/sage/blob/develop/pkgs/sage-conf/sage_conf.py#L1 we have:
`from _sage_conf._conf import *`

`_sage_conf` is an installed top-level package.

Here: `from builddir._conf import *`

`builddir` is not an installed top-level package.

More questions?

Reply:
I don't know what an "installed top-level package" is, but regardless, the editable install works (or do you have a problem with that?!) so that should be good for now. Can be improved in a follow-up.

Reply:
https://docs.python.org/3/reference/import.html

Reply:
What documented behavior is not working for you? As this is a temporary package anyway, I will not invest more time in it than necessary.

Reply:
Calling something "temporary" does not make it exempt from normal scrutiny in a review.
In particular, when no plan has been made known anywhere.

Reply:
Can you then please at least say what is not working for you, and why you need this?

Reply:
Tobias, I said it clearly in the first message. 

The lines `from builddir._conf import *` are faulty because they depend on "." being in PYTHONPATH etc. This is obviously never correct in library code.

The command `sage-config` must work no matter the current working directory.

Reply:
There is no sage-config script in the meson version. Please test it yourself and report back if you encounter any issues.

Reply:
Providing a sage-config script is part of the contract of sage-conf.
So that's another work item.

Reply:
The meson variant is a minimal version of sage-conf that is good enough for an editable install. That's the documented behavior. Please don't extend the scope of this PR by your wishes. If you have a problem that it is only implements a subset of the sage-confirm package, I can rename it. Suggestions welcome

Reply:
Something of this quality is obviously not mergeable. So please refrain from setting it back to "needs review".

Reply:
Always a pleasure to work with you. Not sure how that's possible but I'm still surprised by your flexibility to accept solutions that are slightly different from how you envisioned them.

So you really think without the sage-conf script the Sage world will end? There is no way that this script could be added later when other devs are missing it?

Reply:
https://github.com/sagemath/sage/blob/develop/CODE_OF_CONDUCT.md

Reply:
if this is working it can be merged.

more features may be handled by a follow-up

Reply:
Thanks @dimpase!

Reply:
Note that @dimpase said "if this is working".

I've already made you aware that it does not work. It is not ready to even be tested yet because the defect is obvious from the source code.

Reply:
Matthias, what's your problem? Are you triggered because this PR replaces the configure scripts that you put so much time into? Do you want to pay back because I criticized some of your PRs? You have a lot of things going on in your 'real' life and that stresses you? Whatever it is, please take a deep breath before you further comment here. I'm working on sage in my free time, and if I have to deal with too much negativity I'll just stop. 

Let me tell you a story: A few years ago we had a new contributor at JabRef. He was opening a lot of PRs with huge refactorings. Generally, it was good work but had a few rough edges. At that time, I knew almost any line of the project by heart and thus felt justified to propose alternatives / improvements to these PRs. But this new contributor had also quite strong opinions about how to do things correctly, so we had quite a few long discussions about many things - things which in retrospect were more bike-shedding than actual important issues. At some point, in the middle of one of these PR discussions, he just disappeared. Only at that point, I realized how much good work was in these PRs and that merging 'slightly suboptimal' PRs is still better than having no PRs at all (seems to be a human thing to only realize what you have until you don't have it anymore). It was a good learning experience for me (and I still sometimes have to remind me about it), so maybe this story also has some takeaway points for you.

So that being said, `pip install --no-build-isolation -v -v --editable ./pkgs/sage-conf_meson` produces
```
...
Preparing editable metadata (pyproject.toml) ... done
....
Building wheels for collected packages: sage-conf
....
Building editable for sage-conf (pyproject.toml) ... done
  Created wheel for sage-conf: filename=sage_conf-10.2b7-0.editable-py3-none-any.whl size=2512 sha256=4958feb00d99c295c3e0aa500f37d6197d982b7bbefb7d81c2fe318e5e3f65da
  Stored in directory: /tmp/pip-ephem-wheel-cache-wkm2csaz/wheels/5c/ba/39/add681a35d9ceb3083cbf17c0ed691dcb3688722ca75fb2441
Successfully built sage-conf
```
So could you please test it and report what is not working for you, in the ideal case with suggestions on how to improve things. Thanks!

Reply:
Tobias, this is highly inappropriate. I'm taking this to sage-abuse.

Reply:
Tobias, I'm writing on behalf of the sage-abuse committee.   We request that you remove your [comment](https://github.com/sagemath/sage/pull/36489#discussion_r1372452493) from this thread and replace it by one that is consistent with our code of conduct [1], or request that an admin delete it your comment.    Your comment violates points of the code of conduct.  If you feel that anybody else's behavior violates the code of conduct, we encourage you to email [2] and explain why.   We also encourage you to write whatever rant you want to [3].     -- Thank you.

[1] https://github.com/sagemath/sage/blob/develop/CODE_OF_CONDUCT.md
[2] https://groups.google.com/g/sage-abuse
[3] https://groups.google.com/g/sage-flame

Reply:
@williamstein I honestly don't see where I've violated the code of conduct. I merely (wanted to) express that I don't enjoy Matthias negativity (and were genuinely asking where this negativity is coming from). Anyway, I've now hidden my comment.

@mkoeppe Would you please review the PR in line with the code of conduct? In particular, finally explain to me what errors you encounter. I get: 

`pip install --no-build-isolation -v -v --editable ./pkgs/sage-conf_meson` produces
```
...
Preparing editable metadata (pyproject.toml) ... done
....
Building wheels for collected packages: sage-conf
....
Building editable for sage-conf (pyproject.toml) ... done
  Created wheel for sage-conf: filename=sage_conf-10.2b7-0.editable-py3-none-any.whl size=2512 sha256=4958feb00d99c295c3e0aa500f37d6197d982b7bbefb7d81c2fe318e5e3f65da
  Stored in directory: /tmp/pip-ephem-wheel-cache-wkm2csaz/wheels/5c/ba/39/add681a35d9ceb3083cbf17c0ed691dcb3688722ca75fb2441
Successfully built sage-conf
```
So could you please test it and report what is not working for you, in the ideal case with suggestions on how to improve things. Thanks!

Reply:
I've gone ahead and just deleted the comment (as an admin).  The code says basically be "friendly, patient, welcoming, considerate, respectful and polite".  Let's all do that and move on. 

Look I'm incredibly appreciative and amazed that all of you are contributing to Sage, almost exactly 20 years after I started the project.  Please, please keep doing so, and in a friendly manner.  Thanks!!

Reply:
> Note that @dimpase said "if this is working".
> 
> I've already made you aware that it does not work. It is not ready to even be tested yet because the defect is obvious from the source code.
> 

Please provide a step by step reproducer for the error you see.

Reply:
There's only one step to reproduce: `import sage_conf`.

Reply:
Right, this is indeed not working well. But with the build passing without issues, this means we don't need the sage conf package at all with the conda install?! So we just don't install it at all, and everything is fine?

Reply:
sage-conf is a declared install-requires of sagelib, so some version of it needs to be installed. 
This is all explained in a fairly detailed way in https://github.com/sagemath/sage/blob/develop/pkgs/sage-conf/README.rst, do let me know if anything is missing there and we can improve this documentation.

At this time, I don't have a comment on what variables provided by `sage_conf` are needed at the runtime of the Sage library. We have multiple fallbacks for configuration -- the defaults set in `sage.env`, runtime detection via `sage.features`, and possibly more in various parts of the Sage library.

Reply:
> At this time, I don't have a comment on what variables provided by sage_conf are needed at the runtime of the Sage library.

That's my observation: apparently none. The fallbacks are good enough for conda.

>  sage-conf is a declared install-requires of sagelib, so some version of it needs to be installed.

So can I change that to an optional dependency? (since it seems to be optional)

Reply:
> > At this time, I don't have a comment on what variables provided by sage_conf are needed at the runtime of the Sage library.
> 
> That's my observation: apparently none. The fallbacks are good enough for conda.

You may be right about this, but a lot currently still depends on the assumption "$SAGE_LOCAL = $CONDA_PREFIX" (e.g., see #30914 for pointers), which is guaranteed by the all-conda installation instructions, but I am not sure that we will want to keep it in the future.

A problem with this assumption is that we currently have an inflexible either-or situation: 
- We either take all packages from conda and do not use the Sage build system at all, for anything; 
- or we use the Sage build system for everything and then cannot use Python packages from conda. (For the purpose of this discussion, let's ignore the `--enable-system-site-packages` switch.)

Reply:
> > sage-conf is a declared install-requires of sagelib, so some version of it needs to be installed.
> 
> So can I change that to an optional dependency? (since it seems to be optional)

No, by design it's required, but as is explained in https://github.com/sagemath/sage/blob/develop/pkgs/sage-conf/README.rst, you can pick your own implementation.

Reply:
> > > At this time, I don't have a comment on what variables provided by sage_conf are needed at the runtime of the Sage library.
> > 
> > 
> > That's my observation: apparently none. The fallbacks are good enough for conda.
> 
> You may be right about this, but a lot currently still depends on the assumption "$SAGE_LOCAL = $CONDA_PREFIX" (e.g., see #30914 for pointers), which is guaranteed by the all-conda installation instructions, but I am not sure that we will want to keep it in the future.
> 
> A problem with this assumption is that we currently have an inflexible either-or situation:
> 
>     * We either take all packages from conda and do not use the Sage build system at all, for anything;
> 
>     * or we use the Sage build system for everything and then cannot use Python packages from conda. (For the purpose of this discussion, let's ignore the `--enable-system-site-packages` switch.)

Right, but in the "we take everything from conda" the assumption will always be valid, right? So in this case we don't need a sage_conf package. If in the future some semi-mixed versions are supported, then this method needs its own sage_conf package.

> > So can I change that to an optional dependency? (since it seems to be optional)
> 
> No, by design it's required, but as is explained in https://github.com/sagemath/sage/blob/develop/pkgs/sage-conf/README.rst, you can pick your own implementation.

This document doesn't say anything on why we need the package nor why the design cannot be changed.

Reply:
Either way, I've fixed the package now. Please give it a try.

Reply:
> Right, but in the "we take everything from conda" the assumption will always be valid, right? 

My point is that this dichotomy of the two installation methods is not sustainable; so it would be unwise to be invested in a "pure" mode that seeks to eliminate sage_conf altogether.